IdentityCore/src/cache/mac/MSIDMacCredentialStorageItem.m

```diff
@@ -0,0 +1,217 @@
+    {
+        self.cacheObjects = [NSMutableDictionary dictionary];
+        NSString *queueName = [NSString stringWithFormat:@"com.microsoft.universalstorage-%@", [NSUUID UUID].UUIDString];
+        self.queue = dispatch_queue_create([queueName cStringUsingEncoding:NSASCIIStringEncoding], DISPATCH_QUEUE_CONCURRENT);
+    }
+    
+    return self;
+}
+
+- (void)storeCredential:(MSIDCredentialCacheItem *)credential forKey:(MSIDDefaultCredentialCacheKey *)key
+{
+    dispatch_barrier_async(self.queue, ^{
+        NSString *credentialKey = [self getCredentialKey:key];
+        [self.cacheObjects setObject:credential forKey:credentialKey];
+    });
+}
+
+- (void)mergeStorageItem:(MSIDMacCredentialStorageItem *)storageItem
+{
+    dispatch_barrier_async(self.queue, ^{
+        for (NSString *key in storageItem.cacheObjects)
+        {
+            MSIDCredentialCacheItem *credential = [storageItem.cacheObjects objectForKey:key];
+            if (credential)
+            {
+                [self.cacheObjects setObject:credential forKey:key];
+            }
+            else
+            {
+                MSID_LOG_WITH_CTX_PII(MSIDLogLevelError, nil, @"Credential is nil for key %@ while merging storage credentials.", MSID_PII_LOG_MASKABLE(key));
+            }
+        }
+    });
+}
+
+- (void)removeStoredCredentialForKey:(MSIDDefaultCredentialCacheKey *)key
+{
+    dispatch_barrier_async(self.queue, ^{
+        NSString *credentialKey = [self getCredentialKey:key];
+        MSID_LOG_WITH_CTX_PII(MSIDLogLevelInfo, nil, @"Removing credential for key %@.", MSID_PII_LOG_MASKABLE(credentialKey));
+        [self.cacheObjects removeObjectForKey:credentialKey];
+    });
+}
+
+- (NSArray<MSIDCredentialCacheItem *> *)storedCredentialsForKey:(MSIDDefaultCredentialCacheKey *)key
+{
+    __block NSArray<MSIDCredentialCacheItem *> *credentials =  @[];
+    
+    dispatch_sync(self.queue, ^{
+        if (key.account && key.service)
+        {
+            NSString *credentialKey = [self getCredentialKey:key];
+            MSIDCredentialCacheItem *credential = [self.cacheObjects objectForKey:credentialKey];
+            if (credential)
+            {
+                credentials = @[credential];
+            }
+        }
+        else
+        {
+            NSArray *storedCredentials = [self.cacheObjects allValues];
+            credentials = [storedCredentials filteredArrayUsingPredicate:[self createPredicateForKey:key]];
+        }
+    });
+    
+    return credentials;
+}
+
+/*
+ This api is thread safe only if an immutable object is passed as parameter.
+ */
+- (instancetype)initWithJSONDictionary:(NSDictionary *)json
+                                 error:(NSError * __autoreleasing *)error
+{
+    MSIDMacCredentialStorageItem *instance = [self init];
+    
+    if (instance)
+    {
+        for (NSString *credentialKey in json)
+        {
+            NSDictionary *rtDict = [json msidObjectForKey:credentialKey ofClass:[NSDictionary class]];
```

Comment:
nit: can you log if we get unexpected input (so when rtDict is not of NSDictionary type)? We want to detect cases when someone is writing corrupt data into the keychain easily. Also, change name from rtDict to just dict or credentialDict?